### ISSUE TITLE
BinDeps.PackageManager failed to satisfy dependency libgdal on 32bit Windows and Linux

### ISSUE THREAD

**Issue by visr** [visr/GDAL.jl#42]
Ever since 8087fea74eca384be44fc1b537a8aa570fc341f6 AppVeyor 32 bit builds have been failing with:
```
LoadError: Provider BinDeps.PackageManager failed to satisfy dependency libgdal
while loading C:\Users\appveyor\.julia\v0.6\GDAL\deps\build.jl, in expression starting on line 33
```
Example of full AppVeyor log [here](https://ci.appveyor.com/project/visr/gdal-jl/build/1.0.63/job/amchjtlw8i5pk7gg#L84).

Don't know what is causing this, and don't have a 32 bit Windows at hand to test. Don't know if Linux 32 bit builds are working, or if this is only for Windows.

**Comment by claireh93** [comment]
I also seem to be having this problem when I try to build GDAL on a 64 bit linux machine running Ubuntu 14.04:
```julia
Warning: 'conda-forge' already in 'channels' list, moving to the top
Fetching package metadata .............
Solving package specifications: .

# All requested packages already installed.
# packages in environment at /home/claireh/.julia/v0.6/Conda/deps/usr:
#
gdal                      2.2.3                    py27_0    conda-forge
========================================================[ ERROR: GDAL ]========================================================

LoadError: Provider BinDeps.PackageManager failed to satisfy dependency libgdal
while loading /home/claireh/.julia/v0.6/GDAL/deps/build.jl, in expression starting on line 36

===============================================================================================================================

=======================================================[ BUILD ERRORS ]========================================================

WARNING: GDAL had build errors.

 - packages with build errors remain installed in /home/claireh/.julia/v0.6
 - build the package(s) and all dependencies with `Pkg.build("GDAL")`
 - build a single package by running its `deps/build.jl` script

===============================================================================================================================
```
When I look in the Conda library, libgdal seems to be there, but Julia doesn't seem to find it? Any advice you could give would be much appreciated!

**Comment by visr** [comment]
Hmm I'm not sure, I've seen the same issue on Travis, so something is wrong.
It seems indeed like the conda installation is fine but then it cannot find it. Does it work if you start julia as such:
`LD_LIBRARY_PATH="/home/claireh/.julia/v0.6/Conda/deps/usr/lib/" julia`
And then `Pkg.test("GDAL")`?

And do you have a `home/claireh/.julia/v0.6/GDAL/deps/deps.jl`? What does it say under load dependencies?

**Comment by visr** [comment]
Ok thanks, so the issue is that `libpng` under Conda.jl (this is used by GDAL to read/write PNG) picks up your system `libz`, which is not the version it needs. Inside the `Conda.LIBDIR` there is a version of `libz` that does work.

See more people dealing with this and offering workarounds here:
https://github.com/JuliaPy/PyPlot.jl/issues/151
https://github.com/JuliaPy/Conda.jl/issues/95

This is a short snippet to test if it can load `libgdal`:
```julia
using Conda; Libdl.dlopen(joinpath(Conda.LIBDIR, "libgdal"))
```

As you can see from the PyPlot issue, it is a quite old issue. It might be best resolved by switching from BinDeps to https://github.com/JuliaPackaging/BinaryBuilder.jl sometime soon.

**Comment by evetion** [comment]
BinaryBuilder will not solve this issue. The underlying issue is that the libz shared library is referenced both by Julia itself on startup(!) and indirectly by libgdal via libpng. Therefore, the required, newer libz will never be loaded if it's not done on startup via LD_PRELOAD or LD_LIBRARY_PATH.

**Comment by visr** [comment]
I see. What are the options to fix this?
Here a short overview of versions:

- zlib current release zlib 1.2.11 (15 Jan 2017)
- libpng requirement: zlib 1.0.4 (24 Jul 96) or later
- Julia 0.6 and master (Win64) includes zlib 1.2.8 (28 Apr 2013)

And from the error message above, the libgdal we get is linked to a libpng that expects 1.2.9 (or later?).

So if by controlling the build ourselves we can make sure that libpng gets linked to the same zlib version as is linked to Julia this would be fixed, right? I'm just not sure how to do that if system zlib is used.

For Julia, on Linux the system zlib seems to be used, which is 1.2.8 on the Ubuntu 14.04 (=Travis), creating the issue above. For Julia on Windows we rely on [mingw64-zlib](https://build.opensuse.org/package/show/windows:mingw:win64/mingw64-zlib) which is also still on 1.2.8.

**Comment by visr** [comment]
Hmm when trying to use the 1.2.8 system libz on the Ubuntu 14.04, GDAL doesn't like it:

```
./configure --with-libz=/usr/lib/x86_64-linux-gnu | grep libz
using internal libz code as deflateInit_ from /usr/lib/x86_64-linux-gnu is missing
```
Checks done here:
https://github.com/OSGeo/gdal/commit/c4ef887114c89cef7c4ca1fe15df1a38868ff272

**Comment by visr** [comment]
This should be closed by #46. Note that in that PR I also set the minimum required julia version to 0.6.3, because this version dropped the zlib requirement for julia. (https://github.com/JuliaLang/julia/commit/199a204615abe90e356763242bde3df65391dd69)
